### DISCUSSION TITLE
Design approaches for Passive Components

### DISCUSSION THREAD

**Discussion by SMorettini** [nasa/fprime#1556] [Q&A]
When do you use a **Fprime passive component** and when a normal **C++ library**? How you define the boundaries to decide which solution to use? 

**Example:**

Component A reads some data from some sensors
Component A needs to process them with some algorithm. 
The execution time of the processing algorithm is enough low to do it in a synchronous approach. 

Should the component that reads the data process them using a C++ library containing the algorithm for processing or should it send the data to a passive Fprime component?

---

P.S.: I'm pretty sure both solutions are acceptable but I'm curious to see how other teams define these boundaries.

**Comment by ThibFrgsGmz** [comment]
I encourage you to read the excellent books below:
1- [Clean Architecture](https://www.amazon.fr/Clean-Architecture-Craftsmans-Software-Structure/dp/0134494164) by Uncle Bob
2- [Software Architecture: The Hard Parts: Modern Trade-Off Analyses for Distributed Architectures](https://www.amazon.com/Software-Architecture-Trade-Off-Distributed-Architectures/dp/1492086894) by Neal Ford and al.

It is precisely one of the responsibilities of the software architect to establish the tradeoffs associated with this boundary.
You concede to coupling to the framework to some extent, but at what cost.

Specifically, you'll find the section "Frameworks are tools, not ways of life" on page 202 of Uncle Bob's book.
According to him, your software system should be as independent of any framework as possible.

In Neal Ford's book, his team suggests that it is better to use libraries than to rely on a framework.

You can also read my publication ["Space Flight Software Systems: An approach to understanding their Open Source Framework Paradigm"](https://www.researchgate.net/publication/359602735_Space_Flight_Software_Systems_An_approach_to_understanding_their_Open_Source_Framework_Paradigm) where I discuss this topic and gather references to many sources of information.

**Reply by Joshua-Anderson** [reply]
Fully agree with your points @ThibFrgsGmz. The item to consider when deciding between a component and a library is reusability. If you develop a F' component for a generic sensor, you can fairly easily reuse it between vastly different deployments using fprime, as long as you replace the sensor bus component with a platform specific version.
